### ISSUE TITLE
[question] Use release dependencies with a debug build for the application with CMakeDeps/Toolchain

### ISSUE THREAD

**Issue by jjcasmar** [conan-io/conan#12656]
I am trying to understand how to properly use conan for managing the dependencies of my application. When compiling my application with Release config, everything works fine. The issue comes when I try to compile my application in Debug mode.

My application depends on HDF5, and conan generates a HDF5Config.cmake and HDF5-Target-Release.cmake files (plus some more) when using the default profile. This works fine if I compile my application in Release mode, since it will search for HDF5-Target-Release.cmake file, but fails when compiling my application in Debug mode, as it will search for HDF5-Target-Debug.cmake file, which doesn't exist.

In order to fix this, I need to specify a Debug profile, which in turn will compile my dependencies in Debug mode and generate the HDF5-Target-Debug.cmake.

However, this is a nogo, I want to debug my application, not my dependencies. How should I ask conan to generate to compile my dependencies in Release mode and generate the correct FooConfig.cmake files to be usable in a Debug build of the application?

My conanfile.py looks like this

```
from conan import ConanFile
from conan.tools.cmake import CMakeDeps, CMakeToolchain, cmake_layout

class HDF5Test(ConanFile):
    settings = "os", "arch", "compiler", "build_type"

    def requirements(self):
        self.requires("hdf5/1.13.1")

    def generate(self):
        cmake = CMakeDeps(self)
        cmake.generate()

        toolchain = CMakeToolchain(self)
        toolchain.generate()

    def layout(self):
        cmake_layout(self)
```

- [x] I've read the [CONTRIBUTING guide](https://github.com/conan-io/conan/blob/develop/.github/CONTRIBUTING.md).

**Comment by jcar87** [comment]
Hi @jjcasmar, thanks for asking this question.

>  This works fine if I compile my application in Release mode, since it will search for HDF5-Target-Release.cmake file, but fails when compiling my application in Debug mode, as it will search for HDF5-Target-Debug.cmake file, which doesn't exist.

What error are you getting? It should possible to configure a consumer CMake project in the Debug configuration after having only installed a different configuration (e.g. Release) using Conan. There may still be an error down the line at build time, but the lack fo `HDF5-Target-Debug.cmake` in itself should not cause any errors.


> I want to debug my application, not my dependencies. How should I ask conan to generate to compile my dependencies in Release mode and generate the correct FooConfig.cmake files to be usable in a Debug build of the application?

As of Conan 1.55 (released last week) this should be possible defining the `CMAKE_MAP_IMPORTED_CONFIG_xxx` variable in your consumer CMake project.  For example, if you do:

```
set(CMAKE_MAP_IMPORTED_CONFIG_DEBUG RELEASE)
```

This is a way of telling CMake that when your project is configured in `Debug`, map to the `Release` configuration of the `IMPORTED` targets, if they don't have a Debug configuration. So it should be possible to configure your project in `Debug`, install the release binaries, and have those work if that variable is defined. More information in the docs [here](https://docs.conan.io/en/latest/reference/conanfile/tools/cmake/cmakedeps.html#map-from-project-configuration-to-imported-target-s-configuration).

**Comment by jjcasmar** [comment]
The error I am getting is in compile time for my application. Can't find headers of the dependencies, mainly because it tried to found for Foo-Target-Debug.cmake, which doesn't exist.

Ill try your suggestion. Is is possible to add that variable using the toolchain generated by CMakeToolchain?

**Comment by jjcasmar** [comment]
Or maybe, even easier, could we specify CMakeDeps to generate only the Foo-Target.cmake file, without separating between debug and release config?

**Comment by jcar87** [comment]
Hi @jjcasmar, thanks for the quick reply.

There is already an `HDF5Config.cmake` that doesn't discern between the build type - this one will search for any of the `-target-*.cmake` files and include them _if they exist_.  This is the same behaviour you would find in config and target files generated directly by CMake.

>  Is is possible to add that variable using the toolchain generated by CMakeToolchain?

This is indeed possible. Alternatively, if you want to avoid modifying either the consumers `CMakeLists.txt` or the `conanfile.py`, you can invoke Conan install in such a way that it will generate the files for the debug configuration, pointing at the release files.

```
conan install . -s build_type=Release
conan install . -s &:build_type=Debug -s build_type=Release
```

**Comment by jjcasmar** [comment]
> Hi @jjcasmar, thanks for the quick reply.
> 
> There is already an `HDF5Config.cmake` that doesn't discern between the build type - this one will search for any of the `-target-*.cmake` files and include them _if they exist_. This is the same behaviour you would find in config and target files generated directly by CMake.

I am not absolutely sure about this. If I only compile my dependencies (using `conan install . --build=missing` in Release mode, and I try to compile my app in Debug mode, CMake find_package doesn't fail, but the imported targets are empty and I can't compile my application because it doesn't find the include headers for my dependencies. 

Normally, with CMake you won't have Target-Release or Target-Debug files, only FooConfig, including FooTarget.cmake, and that single file will define the imported target. 

I am actually not sure about this last sentence, but normally I can just use dependencies compiled in Release mode in Debug mode for my application without adding extra variables

**Comment by jcar87** [comment]
> Normally, with CMake you won't have Target-Release or Target-Debug files, only FooConfig, including FooTarget.cmake, and that single file will define the imported target.

CMake will typically add a third file for the configuration that was built. You can see an example [here](https://packages.ubuntu.com/bionic/amd64/libopencv-dev/filelist) -

> I am actually not sure about this last sentence, but normally I can just use dependencies compiled in Release mode in Debug mode for my application without adding extra variables

Indeed that is the case. Currently there is a limitation in CMakeDeps were that is not the current behaviour. In order to avoid adding extra variables, you can try the suggestion above:

```
conan install . -s build_type=Release
conan install . -s &:build_type=Debug -s build_type=Release
```

this will cause CMakeDeps to  generate both sets of files (-Debug and -Release) - but in both cases they will point to the Release configuration of the dependencies, thus achieving the same results.

**Comment by jjcasmar** [comment]
Thanks @jcar87
With the info you gave me I think its enough to fix/workaround this behavior. I think I will stick with adding the variable to the toolchain

**Comment by gouriano** [comment]
@jcar87
Is this syntax official ("documented", in a sense that it will not change in the future)?

    conan install . -s &:build_type=Debug -s build_type=Release

I do not see anything like this in documentation

**Comment by tanqiao-nreal** [comment]
> @jcar87 Is this syntax official ("documented", in a sense that it will not change in the future)?
> 
> ```
> conan install . -s &:build_type=Debug -s build_type=Release
> ```
> 
> I do not see anything like this in documentation

https://github.com/conan-io/conan/issues/13478#issuecomment-1475389368 explain what the "&" means, though I cannot find it in the documentation, too

**Comment by h-2** [comment]
Seeing that there were so many threads about this, we postponed switching to Release-built dependencies, but wanted to tackle it now. To be honest, all the solutions presented here look like "hacks". They are definitely not obvious from the documentation, and they needlessly complicate the build process.

I'd really like to be able to set something like `dependency_build_type=Release` in my profile. In fact, I have never encountered a scenario where I want all dependencies to be built in Debug-mode, so I would even suggest this be the default :)

**Comment by memsharded** [comment]
> I have never encountered a scenario where I want all dependencies to be built in Debug-mode, so I would even suggest this be the default :)

Windows/MSVC will fail to link in most cases if dependencies are not in Debug mode when you are building in Debug. I am afraid that this default cannot be easily changed. We plan some improvements to the ``CMakeDeps`` integration that hopefully might help with this case, but it will take time. In the meantime ``-s build_type=Release -s &:build_type=Debug`` is not a hack, it is a documented and good to use solution

**Comment by h-2** [comment]
Ah, I hadn't considered Windows. It's just that in the Unix world, I had never heard of anyone building all their requirements in Debug mode (unless they wanted to debug a single particular library).

I would still argue that the current behaviour is not expected by many users, and it would be great to have the workaround mentioned more promeninently in the documentation.

**Comment by memsharded** [comment]
Windows user base is a very important one too :) 

Also, Conan try to implement the same default behaviors in all platforms, consistency is important. So ``conan install . -s build_type=Debug`` does the same in all operating systems.

**Comment by poxerial** [comment]
Does conan 1.60.1 support `-s &:build_type=Debug -s build_type=Release`?
I encountered error like this:
```
ninja: error: '/root/.conan/data/protobuf/3.21.4/_/_/package/53f76f058631e1c3a500008c75af3f37d7786131/bin/protoc', needed by '/path/to/my/project/header.h'
```
The path doesn't exist, as we actually didn't intall debug version of dependencies. But it is needed by my project.

**Comment by memsharded** [comment]
Is ``-s &build_type=Debug `` a typo? It should be ``-s &:build_type=Debug``. with the ``:``. I think in theory it should work, but better to update to 1.62, just in case.

This sounds a different thing, about  ``protoc``, which is the executable. It would be better to file a full report as a new ticket.

**Comment by GregDomjan** [comment]
Just a note for myself when I return yet again

When seeing 
> conan install: error: argument -s/--settings: expected one argument
ERROR: Exiting with code: 2

Windows cmd line needs quotes 
` -s "&:build_type=Debug" -s build_type=Release`

**Comment by memsharded** [comment]
Yep, @GregDomjan actually all special characters in the terminal in Linux/Mac (typically not an issue in Windows), need to be quoted too, like ``... -o "*:shared=True"``
